shared-bindings/alarm/SleepMemory.c

```diff
@@ -0,0 +1,180 @@
+/*
+ * This file is part of the Micro Python project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017 Scott Shawcroft for Adafruit Industries
+ * Copyright (c) 2020 Dan Halbert for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/binary.h"
+#include "py/objproperty.h"
+#include "py/runtime.h"
+#include "py/runtime0.h"
+
+#include "shared-bindings/alarm/SleepMemory.h"
+#include "supervisor/shared/translate.h"
+
+//| class SleepMemory:
+//|     """Store raw bytes in RAM that persists during deep sleep.
+//|     The class acts as a ``bytearray``.
+//|     If power is lost, the memory contents are lost.
+//|
```

Comment:
What do you think about having it track the highest written address and having it scale the preserved memory? This would allow for saving larger data such as json. It'd also allow us to only use the registers if only one value is written.

Here in shared-bindings you'd just say, "The len is the total available memory. It may be available over multiple RAM regions. The lowest power memories are at the lowest indices so using the fewest bytes at the start will use the least memory."

Reply:
Based on my discord conversations with @igrr (see transcription below in comment), it sounds better to allocate a fixed chunk of slow or fast RTC mem with `RTC_DATA_ATTR` and have the linker handle mapping the chunk. There is other functionality (ULP, and IDF heap) that wants to use fast or slow RTC memory, and it uses the same linker map mechanism, As I mentioned, the power required for keeping the memory persistent is minimal. Given that we can have ULP alarms too and ULP also needs to use RTC memory also (it can only use slow), so we can use the ESP-IDF linker map mechanisms to keep the storage regions from bumping into each other. So I'm inclined not to try to do this dynamically. There is plenty available and the power cost is low.